src/wrappers/themis/jsthemis/secure_message.cpp

```diff
@@ -45,6 +46,21 @@ namespace jsthemis {
 
   void SecureMessage::New(const Nan::FunctionCallbackInfo<v8::Value>& args) {
     if (args.IsConstructCall()) {
+      if(args.Length()<2){
+        ThrowError("Secure Message constructor", THEMIS_INVALID_PARAMETER, "missing private and public key");
```

Comment:
what about to move these strings with error description to error.h and store them as constants accessible from client's code? it helps to:
* use 1 constant name and value for 1+ usage as a description
* change text in one place for all usages 
* compare error string from client's code and to allow distinguish one error from other (THEMIS_INVALID_PARAMETER was returned for public or private key)

Reply:
Mmm... I'm not really a fan of storing developer-visible human-readable strings in separate files far from their actual usage because as a developer I'd rather see `"missing private and public key"` in the code immediately where the error is reported rather than some cryptic constant `THEMIS_SECURE_MESSAGE_ARGUMENT_COUNT_KEYS`. (Error codes are a different story.)

If I have an error message then grepping for its substrings will lead directly to the source code rather than to the `errors.hpp` where I'd have to do another search for the intermediate constant usage.

I also don't imagine our users _handling_ these errors based on human-readable strings. That's a bad practice. If anything, I'd export all those `themis_status_t` constants and added a proper exception type with some `status` property which then can be programmatically compared to error code constants.

Reply:
I don't like to distinguish error types by string literals too but it's okay to compare some constant values if environment doesn't have any way to use explicit user's exception. as I understand now we return general Exception with some error message - https://github.com/cossacklabs/themis/pull/384/files#diff-55a946083bac69602385c43c0a35e717R83
How user can check in runtime that this error related to incorrect/empty private key and create custom message for user without pass error from our library as is? for example
```
try {
// use secure message with invalid private key
} catch (err) {
  if err.message == jsthemis.INVALID_PRIVATE_KEY {
     alert('you uploaded empty file for private key')
  } 
} 

``` 
instead 
```
try {
// use secure message with invalid private key
} catch (err) {
  alert(err.message); // message == 'secure message: invalid parameter'
} 
```
?

Reply:
Uh... well... currently this is impossible to do consistently with any wrapper in any language. The core library (at best) reports THEMIS_INVALID_PARAMETER and then the user code has to check itself which parameter is invalid, why it is so, and maybe show an appropriate message to the end user in their preferred language.

If machine-readable errors are important I'd suggest to plan a more broad improvement starting from the core: introduce new error codes like THEMIS_INVALID_PRIVATE_KEY, start using them in the core code instead of generic THEMIS_INVALID_PARAMETER, and then update the error reporting across language wrappers. That way we'd be consistent with error reporting.

There's a number of an open questions there. For example, do we need more detailed codes for various reasons why the private key can be invalid? It can be empty, malformed, valid but of unexpected type, does not match the public key, etc.

Reply:
As for this PR, I'd rather not export any new API unless we're sure that it's the right thing and going to support it.

Currently the users of JsThemis do not have an official blessed way to distinguish between Themis errors. However, as developers they may read the documentation, read the error messages they observe during testing, and do something about that on their application level. This PR does not change that fundamentally, but it improves the error messages which makes it a bit easier for developers to guess what's wrong with the way they use Themis.

I think that comparing human-readable error messages is not a good API and we should not start supporting it since 0.11. I'd suggest to keep the current status quo, and maybe introduce a better support for error handling in 0.12+

Reply:
I support @Lagovas – it's important to improve text messages, but it would be nice to allow users to distinguish errors by checking error code / error name as well.

From what I see, it's typical for NaN to throw errors as strings only:
https://github.com/nodejs/nan/blob/master/test/cpp/error.cpp

However can we propagate error number as well? 

Maybe smth similar to

```
  static NAN_INLINE(void NanThrowError(
      const char *msg,
      const int errorNumber)) {
    NanThrowError(NanError(msg, errorNumber));
  }
```

(inspired by http://res.wisedu.com/FS/mockserver/node_modules/nan/nan.h and https://git.visteon.com/external/nan/commit/c1ab008e7d9151a133ad7cd84a8443409e471dcb#7de2fb2c0f5228eb2fbf733d63d9d3203ec08371_350_357)

Reply:
Yeah, it seems that we can use this approach. Then the exception will have an additional property `code` which will contain Themis error code. We should also export error code constants from our module. This should make it possible to handle errors like this:

```js
try {
  // use Themis
}
catch (e) {
  if (e.code == jsthemis.INVALID_PARAMETER) {
    // handle error
  }
}
```